local_planner/src/nodes/histogram.cpp

```diff
@@ -1,106 +1,77 @@
-  resolution = resolution / 2;
-  z_dim = 2 * z_dim;
-  e_dim = 2 * e_dim;
-  std::vector<std::vector<double> > temp_bin;
-  std::vector<std::vector<double> > temp_age;
-  std::vector<std::vector<double> > temp_dist;
-  temp_bin.resize(e_dim);
-  temp_age.resize(e_dim);
-  temp_dist.resize(e_dim);
-  for (int i = 0; i < e_dim; ++i) {
-    temp_bin[i].resize(z_dim);
-    temp_age[i].resize(z_dim);
-    temp_dist[i].resize(z_dim);
+  if (resolution_ != ALPHA_RES * 2) {
+    throw std::logic_error(
+        "Invalid use of function upsample(). This function can only be used on "
+        "a half resolution histogram.");
   }
-  for (int i = 0; i < e_dim; ++i) {
-    for (int j = 0; j < z_dim; ++j) {
+  resolution_ = resolution_ / 2;
+  z_dim_ = 2 * z_dim_;
+  e_dim_ = 2 * e_dim_;
+  Eigen::MatrixXi temp_age(e_dim_, z_dim_);
+  Eigen::MatrixXf temp_dist(e_dim_, z_dim_);
+
+  for (int i = 0; i < e_dim_; ++i) {
+    for (int j = 0; j < z_dim_; ++j) {
       int i_lowres = floor(i / 2);
       int j_lowres = floor(j / 2);
-      temp_bin[i][j] = bin[i_lowres][j_lowres];
-      temp_age[i][j] = age[i_lowres][j_lowres];
-      temp_dist[i][j] = dist[i_lowres][j_lowres];
+      temp_age(i, j) = age_(i_lowres, j_lowres);
+      temp_dist(i, j) = dist_(i_lowres, j_lowres);
     }
   }
-  bin = temp_bin;
-  age = temp_age;
-  dist = temp_dist;
+  age_ = temp_age;
+  dist_ = temp_dist;
 }
 
 void Histogram::downsample() {
-  resolution = 2 * resolution;
-  z_dim = z_dim / 2;
-  e_dim = e_dim / 2;
-  std::vector<std::vector<double> > temp_bin;
-  std::vector<std::vector<double> > temp_age;
-  std::vector<std::vector<double> > temp_dist;
-  temp_bin.resize(e_dim);
-  temp_age.resize(e_dim);
-  temp_dist.resize(e_dim);
-  for (int i = 0; i < e_dim; ++i) {
-    temp_bin[i].resize(z_dim);
-    temp_age[i].resize(z_dim);
-    temp_dist[i].resize(z_dim);
+  if (resolution_ != ALPHA_RES) {
+    throw std::logic_error(
+        "Invalid use of function downsample(). This function can only be used "
+        "on a full resolution histogram.");
   }
-  for (int i = 0; i < e_dim; ++i) {
-    for (int j = 0; j < z_dim; ++j) {
+  resolution_ = 2 * resolution_;
+  z_dim_ = z_dim_ / 2;
+  e_dim_ = e_dim_ / 2;
+  Eigen::MatrixXi temp_age(e_dim_, z_dim_);
+  Eigen::MatrixXf temp_dist(e_dim_, z_dim_);
+
+  for (int i = 0; i < e_dim_; ++i) {
+    for (int j = 0; j < z_dim_; ++j) {
       int i_high_res = 2 * i;
       int j_high_res = 2 * j;
 
-      double mean_bin =
-          (bin[i_high_res][j_high_res] + bin[i_high_res + 1][j_high_res] +
-           bin[i_high_res][j_high_res + 1] +
-           bin[i_high_res + 1][j_high_res + 1]) /
-          4.0;
-      double mean_age =
-          (age[i_high_res][j_high_res] + age[i_high_res + 1][j_high_res] +
-           age[i_high_res][j_high_res + 1] +
-           age[i_high_res + 1][j_high_res + 1]) /
-          4.0;
+      float mean_age =
+          (age_(i_high_res, j_high_res) + age_(i_high_res + 1, j_high_res) +
```

Comment:
This can be simplified even further with
`temp_age(i, j) = age_.block(i_high_res, j_high_res, 2, 2).mean();`

I <3 Eigen

Reply:
<3